csrc/layernorm_kernels.cu

```diff
@@ -109,12 +151,27 @@ void fused_add_rms_norm(
     input.scalar_type(),
     "fused_add_rms_norm_kernel",
     [&] {
-      vllm::fused_add_rms_norm_kernel<scalar_t><<<grid, block, 0, stream>>>(
+
+      bool use_shmem = true;
+      //estimate the shared memory size
+      int shmem_size = hidden_size * sizeof(scalar_t);
+
+      if (shmem_size > getMaxSharedMemoryPerBlock(input)) {
+        shmem_size = 0;
+        use_shmem = false;
+      }
+      if (shmem_size >=48 * 1024) {
```

Comment:
A quick question: Where does this number (48 * 1024) come from?

Reply:
I read some documents: https://on-demand.gputechconf.com/gtc-il/2018/pdf/sil8140-optimizing-cuda-applications-for-the-volta-turing-gpu-architecture.pdf
> To overcome 48 KB per threadblock limitation call:
>  cudaFuncSetAttribute (MyKernel, cudaFuncAttributeMaxDynamicSharedMemorySize, maxsize);

In my own test. if I allocated more than 48KB shared memory without this 'SetAttribute', the cuda runtime will panic.

and it is easy to find NVIDIA's own library has similar code:

https://github.com/NVIDIA/cutlass/blob/47a3ebbea9860e14c095b52c4e6e2db33340f572/include/cutlass/gemm/device/ell_gemm.h#L510